### PR TITLE
Refactor(core): Remove top level `ɵdefine*` for better tree shaking support.

### PR DESCRIPTION
**Pull request by mhevery**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: packages/core/src/render3/definition.ts [hunk 1]
```diff
@@ -700,32 +700,28 @@ export const ɵɵdefineDirective =
...
-           name: pipeDef.name,
-           factory: null,
-           pure: pipeDef.pure !== false,
-           onDestroy: pipeDef.type.prototype.ngOnDestroy || null
-         }) as never;
+/**
+ * Type representing all of the optional definition properties which could be placed on the type.
+ */
+export interface DefType<T> extends Type<T> {
+  [NG_PIPE_DEF]?: PipeDef<T>|ProtoPipeDef<T>;
```
Comment (IgorMinar) [inline packages/core/src/render3/definition.ts:724]: just PipeDef? what about the rest?

File: packages/core/src/render3/definition.ts [hunk 2]
```diff
@@ -742,8 +738,10 @@ export function getDirectiveDef<T>(type: any): DirectiveDef<T>|null {
   return type[NG_DIR_DEF] || null;
 }
 
-export function getPipeDef<T>(type: any): PipeDef<T>|null {
-  return type[NG_PIPE_DEF] || null;
+export function getPipeDef<T>(type: DefType<any>): PipeDef<T>|null {
+  const value = type[NG_PIPE_DEF] || null;
+  return value == null || value instanceof PipeDef ? value :
```
Comment (IgorMinar) [inline packages/core/src/render3/definition.ts:743]: I'd break this up into an `if` followed up by `return value;` to make the code easier to read.

File: packages/core/src/render3/definition.ts (hunk 1 above)
Comment (IgorMinar) [reply packages/core/src/render3/definition.ts:724]: I guess this is just WIP

**Review by IgorMinar — COMMENTED** [review]
This seems to be on the right track, assuming that you'll also update the compiler to emit the code without definePipe fn call.

File: packages/core/src/render3/interfaces/definition.ts [hunk 3]
```diff
@@ -412,6 +428,14 @@ export interface PipeDef<T> {
 
   /* The following are lifecycle hooks for this pipe */
   onDestroy: (() => void)|null;
+
+  constructor(protoPipDef: ProtoPipeDef<T>) {
```
Comment (alxhub) [inline packages/core/src/render3/interfaces/definition.ts:432]:
```suggestion
  constructor(protoPipeDef: ProtoPipeDef<T>) {
```

File: packages/core/src/render3/interfaces/definition.ts [hunk 4]
```diff
@@ -412,6 +428,14 @@ export interface PipeDef<T> {
...
+    this.type = protoPipDef.type;
+    this.name = protoPipDef.name;
+    this.factory = null;
+    this.pure = protoPipDef.pure !== false;
+    this.onDestroy = protoPipDef.type.prototype.ngOnDestroy || null;
```
Comment (alxhub) [inline packages/core/src/render3/interfaces/definition.ts:437]: Do we prefer `||` or `??` for this kind of "X or default value" operation?

File: packages/core/src/render3/definition.ts (hunk 2 above)
Comment (alxhub) [reply packages/core/src/render3/definition.ts:743]: Agreed, this is a bit too compact for me (though I understand why we want to minimize code size in this particular case)

**Comment by alxhub** [comment]
We should probably also change the type of `PipeDefWithMeta` to be `unknown`.

**Pull request closed without merging**